"collecting" output with --color=yes in Continuous Integration output

Hi,

We like to use `--color=yes` when running py.test in continuous integration services like Jenkins (with the [Ansi Color Plugin](https://wiki.jenkins-ci.org/display/JENKINS/AnsiColor+Plugin)). This enables us to see the same nice colors in Jenkins that we see on the terminal.

The only problem is that the "collecting" display progress actually piles up in the terminal because py.test output is being redirected and captured by the server. This can easily be reproduced by redirecting to a file instead:

```python
import pytest
@pytest.mark.parametrize('i', range(10))
def test_foo(i):
    if i == 5:
        assert 0
```

```
py.test --color=yes test_foo.py > out
```

Contents of `out`:

```
[1m============================= test session starts =============================[0m
platform win32 -- Python 2.7.11, pytest-2.8.5, py-1.4.31, pluggy-0.3.1
rootdir: x:\jobs_done10, inifile: 
[1m
collecting 0 items[0m[1m
collecting 10 items[0m[1m
collected 10 items 
[0m
foo.py .....F....

================================== FAILURES ===================================
[1m[31m_________________________________ test_foo[5] _________________________________[0m

i = 5

[1m    @pytest.mark.parametrize('i', range(10))[0m
[1m    def test_foo(i):[0m
[1m        if i == 5:[0m
[1m>           assert 0[0m
[1m[31mE           assert 0[0m

foo.py:6: AssertionError
[1m[31m===================== 1 failed, 9 passed in 0.02 seconds ======================[0m
```

The color codes are correct, but the "collecting" messages are a problem because they occupy many lines of output in test suites with hundreds of tests.

Is there any way to prevent those "collecting" messages from appearing? `-q` gets rid of them, but I want the header to appear, specially in CI.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -72,7 +72,9 @@
 
 *
 
-*
+* Collection only displays progress ("collecting X items") when in a terminal.
+  This avoids cluttering the output when using ``--color=yes`` to obtain
+  colors in CI integrations systems (`#1397`_).
 
 *
 
@@ -106,6 +108,7 @@
 .. _#1226: https://github.com/pytest-dev/pytest/pull/1226
 .. _#1290: https://github.com/pytest-dev/pytest/pull/1290
 .. _#1355: https://github.com/pytest-dev/pytest/pull/1355
+.. _#1397: https://github.com/pytest-dev/pytest/issues/1397
 .. _@biern: https://github.com/biern
 .. _@MichaelAquilina: https://github.com/MichaelAquilina
 .. _@bukzor: https://github.com/bukzor

--- a/_pytest/terminal.py
+++ b/_pytest/terminal.py
@@ -112,6 +112,7 @@ class TerminalReporter:
         self.currentfspath = None
         self.reportchars = getreportopt(config)
         self.hasmarkup = self._tw.hasmarkup
+        self.isatty = file.isatty()
 
     def hasopt(self, char):
         char = {'xfailed': 'x', 'skipped': 's'}.get(char, char)
@@ -234,7 +235,7 @@ class TerminalReporter:
                 self.currentfspath = -2
 
     def pytest_collection(self):
-        if not self.hasmarkup and self.config.option.verbose >= 1:
+        if not self.isatty and self.config.option.verbose >= 1:
             self.write("collecting ... ", bold=True)
 
     def pytest_collectreport(self, report):
@@ -244,7 +245,7 @@ class TerminalReporter:
             self.stats.setdefault("skipped", []).append(report)
         items = [x for x in report.result if isinstance(x, pytest.Item)]
         self._numcollected += len(items)
-        if self.hasmarkup:
+        if self.isatty:
             #self.write_fspath_result(report.nodeid, 'E')
             self.report_collect()
 
@@ -263,7 +264,7 @@ class TerminalReporter:
             line += " / %d errors" % errors
         if skipped:
             line += " / %d skipped" % skipped
-        if self.hasmarkup:
+        if self.isatty:
             if final:
                 line += " \n"
             self.rewrite(line, bold=True)

--- a/testing/test_terminal.py
+++ b/testing/test_terminal.py
@@ -561,11 +561,36 @@ def test_color_yes(testdir):
     assert 'test session starts' in result.stdout.str()
     assert '\x1b[1m' in result.stdout.str()
 
+
 def test_color_no(testdir):
     testdir.makepyfile("def test_this(): assert 1")
     result = testdir.runpytest('--color=no')
     assert 'test session starts' in result.stdout.str()
     assert '\x1b[1m' not in result.stdout.str()
+
+
+@pytest.mark.parametrize('verbose', [True, False])
+def test_color_yes_collection_on_non_atty(testdir, verbose):
+    """skip collect progress report when working on non-terminals.
+    #1397
+    """
+    testdir.makepyfile("""
+        import pytest
+        @pytest.mark.parametrize('i', range(10))
+        def test_this(i):
+            assert 1
+    """)
+    args = ['--color=yes']
+    if verbose:
+        args.append('-vv')
+    result = testdir.runpytest(*args)
+    assert 'test session starts' in result.stdout.str()
+    assert '\x1b[1m' in result.stdout.str()
+    assert 'collecting 10 items' not in result.stdout.str()
+    if verbose:
+        assert 'collecting ...' in result.stdout.str()
+    assert 'collected 10 items' in result.stdout.str()
+
 
 def test_getreportopt():
     class config:

